pkg/upgrademonitor/upgrade_monitor.go

```diff
@@ -0,0 +1,201 @@
+package upgrademonitor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	machineconfigurationalphav1 "github.com/openshift/client-go/machineconfiguration/applyconfigurations/machineconfiguration/v1alpha1"
+	mcfgclientset "github.com/openshift/client-go/machineconfiguration/clientset/versioned"
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	v1 "github.com/openshift/api/config/v1"
+	mcfgalphav1 "github.com/openshift/api/machineconfiguration/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+type Condition struct {
+	State   mcfgalphav1.StateProgress
+	Reason  string
+	Message string
+}
+
+func GenerateAndApplyMachineConfigNodes(ParentCondition *Condition, ChildCondition *Condition, ParentStatus metav1.ConditionStatus, ChildStatus metav1.ConditionStatus, node *corev1.Node, mcfgClient mcfgclientset.Interface, fgAccessor featuregates.FeatureGateAccess) error {
```

Comment:
This function is quite complicated :) mind breaking it down?

Reply:
I will add some comments :)